File: 3762.minimum-operations-to-equalize-subarrays.go
#
# @lc app=leetcode id=3762 lang=golang
#
# [3762] Minimum Operations to Equalize Subarrays
#

# @lc code=start
import (
	"math"
	"sort"
)

func minOperations(nums []int, k int, queries [][]int) []int64 {
	n := len(nums)

	if n == 0 {
		return []int64{}
	}

	q := len(queries)

	pos := make([]int64, n)
	mods := make([]int, n)

	for i := 0; i < n; i++ {
		pos[i] = int64(nums[i]) / int64(k)
		mods[i] = nums[i] % k
	}

	all_pos := make([]int64, n)
	copy(all_pos, pos)
	sort.Slice(all_pos, func(i, j int) bool { return all_pos[i] < all_pos[j] })

	unique_pos := make([]int64, 0, n)
	for i := 0; i < n; i++ {
		if i == 0 || all_pos[i] != all_pos[i-1] {
			unique_pos = append(unique_pos, all_pos[i])
		}
	}

	M := len(unique_pos)

	rank := make([]int, n)
	for i := 0; i < n; i++ {
		rank[i] = sort.Search(M, func(j int) bool { return unique_pos[j] >= pos[i] }) + 1
	}

	type Fenwick struct {
		n   int
		tree []int64
	}

	func newFenwick(size int) *Fenwick {
		return &Fenwick{size, make([]int64, size+2)}
	}

	func (f *Fenwick) update(idx int, val int64) {
		for idx <= f.n {
			f.tree[idx] += val
			idx += idx & -idx
		}
	}

	func (f *Fenwick) query(idx int) int64 {
		sum := int64(0)
		for ; idx > 0; idx -= idx & -idx {
			sum += f.tree[idx]
		}
		return sum
	}

	type Query struct {
		idx, l, r int
	}

	qrys := make([]Query, q)
	for i := 0; i < q; i++ {
		qrys[i] = Query{i, queries[i][0], queries[i][1]}
	}

	block_size := int(math.Sqrt(float64(n)))
	if block_size == 0 {
		block_size = 1
	}

	sort.Slice(qrys, func(i, j int) bool {
		bi := qrys[i].l / block_size
		bj := qrys[j].l / block_size
		if bi != bj {
			return bi < bj
		}
		return qrys[i].r < qrys[j].r
	})

	ft_count := newFenwick(M)
	ft_sum := newFenwick(M)

	mod_freq := make(map[int]int)

	current_l := 0
	current_r := -1
	var current_total_sum int64 = 0
	current_distinct := 0

	ans := make([]int64, q)

	add := func(idx int) {
		md := mods[idx]
		mod_freq[md]++
		if mod_freq[md] == 1 {
			current_distinct++
		}
	
		current_total_sum += pos[idx]
	
		rk := rank[idx]
		ft_count.update(rk, 1)
		ft_sum.update(rk, pos[idx])
	}

	remove := func(idx int) {
		md := mods[idx]
		mod_freq[md]--
		if mod_freq[md] == 0 {
			current_distinct--
		}
	
		current_total_sum -= pos[idx]
	
		rk := rank[idx]
		ft_count.update(rk, -1)
		ft_sum.update(rk, -pos[idx])
	}

	get_sum_smallest := func(kk int) int64 {
		if kk == 0 {
			return 0
		}
	
		lo := 1
		hi := M
		for lo < hi {
			mid := (lo + hi) / 2
			if ft_count.query(mid) >= int64(kk) {
				hi = mid
			} else {
				lo = mid + 1
			}
		}
	
		rr := lo
	
		prev_c := ft_count.query(rr - 1)
		sum_prev := ft_sum.query(rr - 1)
		need := int64(kk) - prev_c
		p_rr := unique_pos[rr-1]
		return sum_prev + need * p_rr
	}

	get_kth := func(kk int) int64 {
		lo := 1
		hi := M
		for lo < hi {
			mid := (lo + hi) / 2
			if ft_count.query(mid) >= int64(kk) {
				hi = mid
			} else {
				lo = mid + 1
			}
		}
	
		return unique_pos[lo-1]
	}

	for _, qq := range qrys {
		ql := qq.l
		qr := qq.r

		for current_l > ql {
			current_l--
			add(current_l)
		}

		for current_r < qr {
			current_r++
			add(current_r)
		}

		for current_l < ql {
			remove(current_l)
			current_l++
		}

		for current_r > qr {
			remove(current_r)
			current_r--
		}

		length := current_r - current_l + 1

		if current_distinct == 1 {
			h := length / 2
			sum_bot := get_sum_smallest(h)
			if length%2 == 0 {
				ans[qq.idx] = current_total_sum - 2 * sum_bot
			} else {
				med := get_kth(h + 1)
				ans[qq.idx] = current_total_sum - 2*sum_bot - med
			}
		} else {
			ans[qq.idx] = -1
		}
	}

	return ans
}
# @lc code=end

File: 3762.minimum-operations-to-equalize-subarrays.java
#
# @lc app=leetcode id=3762 lang=java
#
# [3762] Minimum Operations to Equalize Subarrays
#

# @lc code=start
import java.util.*;
class Solution {
    static class Fenwick {
        long[] tree;
        int n;
        Fenwick(int _n) {
            n = _n;
            tree = new long[n + 2];
        }
        void update(int idx, long val) {
            while (idx <= n) {
                tree[idx] += val;
                idx += idx & -idx;
            }
        }
        long query(int idx) {
            long sum = 0;
            while (idx > 0) {
                sum += tree[idx];
                idx -= idx & -idx;
            }
            return sum;
        }
    }

    static class MoQuery {
        int l, r, idx;
        MoQuery(int l, int r, int idx) {
            this.l = l;
            this.r = r;
            this.idx = idx;
        }
    }

    public long[] minOperations(int[] nums, int k, int[][] queries) {
        int n = nums.length;
        int qnum = queries.length;
        int[] mods = new int[n];
        int[] qvals = new int[n];
        for (int i = 0; i < n; i++) {
            mods[i] = nums[i] % k;
            qvals[i] = nums[i] / k;
        }
        // Sparse Table for min/max mods
        int LOG = 16;
        int[][] stmin = new int[LOG][n];
        int[][] stmax = new int[LOG][n];
        for (int i = 0; i < n; i++) {
            stmin[0][i] = stmax[0][i] = mods[i];
        }
        for (int j = 1; j < LOG; j++) {
            for (int i = 0; i + (1 << j) <= n; i++) {
                stmin[j][i] = Math.min(stmin[j - 1][i], stmin[j - 1][i + (1 << (j - 1))]);
                stmax[j][i] = Math.max(stmax[j - 1][i], stmax[j - 1][i + (1 << (j - 1))]);
            }
        }
        // Identify valid queries
        List<MoQuery> validQueries = new ArrayList<>();
        long[] answer = new long[qnum];
        Arrays.fill(answer, -1L);
        for (int qi = 0; qi < qnum; qi++) {
            int L = queries[qi][0];
            int R = queries[qi][1];
            int lenq = R - L + 1;
            int lg = 31 - Integer.numberOfLeadingZeros(lenq);
            int min_mod = Math.min(stmin[lg][L], stmin[lg][R - (1 << lg) + 1]);
            int max_mod = Math.max(stmax[lg][L], stmax[lg][R - (1 << lg) + 1]);
            if (min_mod == max_mod) {
                validQueries.add(new MoQuery(L, R, qi));
            }
        }
        if (validQueries.isEmpty()) {
            return answer;
        }
        // Discretization
        Integer[] qarr = new Integer[n];
        for (int i = 0; i < n; i++) qarr[i] = qvals[i];
        Arrays.sort(qarr);
        List<Integer> uniqueVals = new ArrayList<>();
        if (n > 0) {
            uniqueVals.add(qarr[0]);
            for (int i = 1; i < n; i++) {
                if (!qarr[i].equals(qarr[i - 1])) {
                    uniqueVals.add(qarr[i]);
                }
            }
        }
        int numUnique = uniqueVals.size();
        long[] valOfRank = new long[numUnique + 1];
        Map<Integer, Integer> rankMap = new HashMap<>();
        for (int i = 0; i < numUnique; i++) {
            int v = uniqueVals.get(i);
            int rnk = i + 1;
            rankMap.put(v, rnk);
            valOfRank[rnk] = v;
        }
        int[] ranks = new int[n];
        for (int i = 0; i < n; i++) {
            ranks[i] = rankMap.get(qvals[i]);
        }
        // Mo's setup
        int blockSz = (int) Math.sqrt(n) + 1;
        validQueries.sort((a, b) -> {
            int ba = a.l / blockSz;
            int bb = b.l / blockSz;
            if (ba != bb) return Integer.compare(ba, bb);
            return Integer.compare(a.r, b.r);
        });
        Fenwick countFT = new Fenwick(numUnique + 1);
        Fenwick sumFT = new Fenwick(numUnique + 1);
        int currL = 0;
        int currR = -1;
        for (MoQuery mq : validQueries) {
            int targetL = mq.l;
            int targetR = mq.r;
            while (currL > targetL) {
                currL--;
                int rk = ranks[currL];
                long v = qvals[currL];
                countFT.update(rk, 1);
                sumFT.update(rk, v);
            }
            while (currR < targetR) {
                currR++;
                int rk = ranks[currR];
                long v = qvals[currR];
                countFT.update(rk, 1);
                sumFT.update(rk, v);
            }
            while (currL < targetL) {
                int rk = ranks[currL];
                long v = qvals[currL];
                countFT.update(rk, -1);
                sumFT.update(rk, -v);
                currL++;
            }
            while (currR > targetR) {
                int rk = ranks[currR];
                long v = qvals[currR];
                countFT.update(rk, -1);
                sumFT.update(rk, -v);
                currR--;
            }
            // Compute ops
            int length = targetR - targetL + 1;
            int kth = (length + 1) / 2;
            int leftb = 1, rightb = numUnique;
            int medRank = rightb + 1;
            while (leftb <= rightb) {
                int midR = leftb + (rightb - leftb) / 2;
                if (countFT.query(midR) >= kth) {
                    medRank = midR;
                    rightb = midR - 1;
                } else {
                    leftb = midR + 1;
                }
            }
            long medVal = valOfRank[medRank];
            long countLeft = countFT.query(medRank - 1);
            long sumLeft = sumFT.query(medRank - 1);
            long countLeq = countFT.query(medRank);
            long sumLeq = sumFT.query(medRank);
            long countRight = (long) length - countLeq;
            long sumRight = sumFT.query(numUnique) - sumLeq;
            long ops = countLeft * medVal - sumLeft + sumRight - countRight * medVal;
            answer[mq.idx] = ops;
        }
        return answer;
    }
}
# @lc code=end